File: Striver/Trees/Q26.java
package Striver.Trees;

import java.util.*;

public class Q26 {
    // Print all the nodes at distance k from the given node in a binary tree

    public void markParents(Node root, Map<Node, Node> parents, Node target){
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        while (!q.isEmpty()){
            Node curr = q.poll();
            if (curr.left != null){
                parents.put(curr.left, curr);
                q.add(curr.left);
            }
            if (curr.right != null){
                parents.put(curr.right, curr);
                q.add(curr.right);
            }
        }
    }

    public List<Integer> distanceK(Node root, Node target, int k) {
        Map<Node, Node> parents = new HashMap<>();
        markParents(root, parents, target);

        Queue<Node> q = new LinkedList<>();
        q.add(target);
        Set<Node> seen = new HashSet<>();
        seen.add(target);
        int dist = 0;
        while (!q.isEmpty()){
            if (dist == k){
                List<Integer> ans = new ArrayList<>();
                for (Node node : q){
                    ans.add(node.data);
                }
                return ans;
            }
            int size = q.size();
            for (int i = 0; i < size; i++){
                Node curr = q.poll();
                if (curr.left != null && !seen.contains(curr.left)){
                    seen.add(curr.left);
                    q.add(curr.left);
                }
                if (curr.right != null && !seen.contains(curr.right)){
                    seen.add(curr.right);
                    q.add(curr.right);
                }
                Node parent = parents.get(curr);
                if (parent != null && !seen.contains(parent)){
                    seen.add(parent);
                    q.add(parent);
                }
            }
            dist++;
        }
        return new ArrayList<>();
    }

    public static void main(String[] args) {
        Node root = new Node(3);
        root.left = new Node(5);
        root.right = new Node(1);
        root.left.left = new Node(6);
        root.left.right = new Node(2);
        root.right.left = new Node(0);
        root.right.right = new Node(8);
        root.left.right.left = new Node(7);
        root.left.right.right = new Node(4);

        Q26 ob = new Q26();
        List<Integer> ans = ob.distanceK(root, root.left, 2);
        System.out.println(ans);
    }
}

File: Striver/Trees/Q26.py
# print all the nodes at distance k from the given node in a binary tree

from Node import Node

def markParents(root, parentsMap, target):
    q = []
    q.append(root)
    while len(q) > 0:
        curr = q.pop(0)
        if curr.left:
            parentsMap[curr.left] = curr
            q.append(curr.left)
        if curr.right:
            parentsMap[curr.right] = curr
            q.append(curr.right)

def printKNodes(root, target, k):
    parentsMap = {}
    markParents(root, parentsMap, target)
    q = []
    q.append(target)
    visited = set()
    visited.add(target)
    level = 0
    while len(q) > 0:
        if level == k:
            break
        size = len(q)
        for i in range(size):
            curr = q.pop(0)
            if curr.left and curr.left not in visited:
                q.append(curr.left)
                visited.add(curr.left)
            if curr.right and curr.right not in visited:
                q.append(curr.right)
                visited.add(curr.right)
            if parentsMap[curr] and parentsMap[curr] not in visited:
                q.append(parentsMap[curr])
                visited.add(parentsMap[curr])
        level += 1
    for i in q:
        print(i.data, end=" ")

root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
target = root.left.left
root.left.right = Node(5)
root.right.left = Node(6)
root.right.right = Node(7)
root.right.right.right = Node(8)
printKNodes(root, target, 2)